crates/rune/src/doc/autocomplete.rs

```diff
@@ -0,0 +1,388 @@
+                }
+                Kind::Struct => {
+                    for (_, name) in self.ctx.iter_components(item)? {
+                        let item = item.join([name])?;
+                        self.collect_meta(&item)?;
+                    }
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+                Kind::Variant => {
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+                Kind::Enum => {
+                    for (_, name) in self.ctx.iter_components(item)? {
+                        let item = item.join([name])?;
+                        self.collect_meta(&item)?;
+                    }
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+                Kind::Macro => {
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+                Kind::Function(f) => {
+                    if f.arg_names.into_iter().flatten().next().map(|s| s.as_str()) == Some("self")
+                    {
+                        self.instance.try_insert(item.try_clone()?, meta)?;
+                    } else {
+                        self.fixed.try_insert(item.try_clone()?, meta)?;
+                    }
+                }
+                Kind::Const(_) => {
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+                Kind::Module => {
+                    for (_, name) in self.ctx.iter_components(item)? {
+                        let item = item.join([name])?;
+                        self.collect_meta(&item)?;
+                    }
+                    self.fixed.try_insert(item.try_clone()?, meta)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn build(&mut self, artifacts: &mut Artifacts) -> Result<()> {
+        let mut content = std::string::String::new();
+        self.write(&mut content)?;
+
+        artifacts.asset(false, "autocomplete.js", || {
+            let string = String::try_from(content)?;
+            Ok(string.into_bytes().into())
+        })?;
+
+        Ok(())
+    }
+
+    fn doc_to_html(&self, meta: &Meta) -> Result<Option<String>> {
+        let mut input = String::new();
+
+        for line in meta.docs {
+            let line = line.strip_prefix(' ').unwrap_or(line);
+            input.try_push_str(line)?;
+            input.try_push('\n')?;
+        }
+
+        let mut o = String::new();
+        write!(o, "<div class=\"docs\">")?;
+        let mut options = Options::empty();
+        options.insert(Options::ENABLE_STRIKETHROUGH);
+
+        let iter = Parser::new_ext(&input, options);
+
+        markdown::push_html(&self.syntax_set, &mut o, iter, None)?;
+
+        write!(o, "</div>")?;
+        let o = String::try_from(o.replace('`', "\\`"))?;
+
+        Ok(Some(o))
+    }
+
+    fn get_name(item: &ItemBuf) -> Result<String> {
+        let mut name = item.try_to_string()?;
```

Comment:
Some of the below seems to be using the wrong range, for example `::std::io::println!()` is longer than `12`, but I'm also curious: What is this doing? :sweat_smile:

A better implementation would probably be:

```rust
const PATTERNS: &[&str] = &["foo", "bar"];

for pat in PATTERNS {
    if name.starts_with(pat) {
        name.try_replace_range(..pat.len(), "")?;
    }
}

Ok(name)
```

Reply:
All right, so the build includes a [prefix](https://github.com/rune-rs/rune/blob/main/crates/rune/src/build.rs#L280). It might be possible to make this available somehow. It includes all the things imported by default.

Reply:
Ah perfect! That's exactly what I looked for. Now it works as expected.
From my perspective this PR is now ready to go.